pkg/authn/challenge/challenge.go

```diff
@@ -0,0 +1,118 @@
+package challenge
+
+import (
+	"context"
+	"crypto/rsa"
+	"embed"
+	"encoding/json"
+	"net/http"
+
+	"github.com/bacalhau-project/bacalhau/pkg/authn"
+	"github.com/bacalhau-project/bacalhau/pkg/lib/policy"
+	"github.com/bacalhau-project/bacalhau/pkg/system"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/samber/lo"
+)
+
+//go:embed *.rego
+var policies embed.FS
+
+type policyData struct {
+	SigningKey jwk.Key `json:"signingKey"`
+	NodeID     string  `json:"nodeId"`
+	ClientID   string  `json:"clientId"`
+}
+
+type response struct {
+	PhraseSignature string `json:"phraseSignature"`
+
+	// A base64 encoded PKCS1 public key, as generated by config.encodePublicKey
+	PublicKey string `json:"publicKey"`
+}
+
+type request struct {
+	InputPhrase []byte `json:"inputPhrase"`
+}
+
+type phraser = func() ([]byte, error)
+
+type challengeAuthenticator struct {
+	authnPolicy *policy.Policy
+	getPhrase   phraser
+	key         jwk.Key
+	nodeID      string
+
+	challenge policy.Query[policyData]
+}
+
+func NewAuthenticator(p *policy.Policy, getPhrase phraser, key *rsa.PrivateKey, nodeID string) authn.Authenticator {
+	return challengeAuthenticator{
+		authnPolicy: p,
+		getPhrase:   getPhrase,
+		key:         lo.Must(jwk.New(key)),
+		nodeID:      nodeID,
+		challenge:   policy.AddQuery[policyData](p, authn.PolicyTokenRule),
+	}
+}
+
+// Authenticate implements Authenticator.
+func (authenticator challengeAuthenticator) Authenticate(req *http.Request) (authn.Authentication, error) {
+	var userInput response
+	err := json.NewDecoder(req.Body).Decode(&userInput)
+	if err != nil {
+		return authn.Authentication{Success: false, Reason: err.Error()}, err
```

Comment:
Might it be worth a new constructor for authn.Authentication that takes an err and returns this structure (and the err). Failed authentication seems like a reasonably common occurrence and it would be good to ensure consistency.

Literally something like 

```go
func FailedAuthentication(err error) error {
     return authn.Authentication{Success: false, Reason: err.Error()}, err
}
```

Reply:
Yes, good idea.